Add subgroups field to the user group objects and new api endpoints for different user group operations.

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
- First commit adds `subgroups` field to the user group objects.
- Second commit adds functions in `actions.py` to add and remove subgroups of a user group.
- Third commit adds API endpoint for updating subgroups of a user group.
- Fourth and Fifith commit add helper functions - `get_user_group_member_ids` and `is_user_in_group`.
- Sixth and Seventh commit add new api endpoints to check whether a user is part of a user group or not.
- Last commit adds endpoint to get members of a user group.

Part of #19525.
 <!-- How have you tested? -->

 <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->

<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

sahil839 commented on zerver/views/user_groups.py:
```diff
@@ -237,3 +237,20 @@ def get_user_group_membership(
+    request: HttpRequest,
+    user_profile: UserProfile,
+    user_group_id: int = REQ(json_validator=check_int, path_only=True),
+    recursive: bool = REQ(json_validator=check_bool, default=True),
```
Do we want to name this parameter as something else like `include_subgroup_members` for the API or is this fine?

timabbott replied:
Let's have an "#api design" discussion on this; I could imagine a name more like `include_subgroups`.

Actually, probably best to do one topic about each of the core changes here:
* API for this endpoint.
* Format of user group update events.
* Etc.

sahil839 commented on templates/zerver/help/include/rest-endpoints.md:
```diff
@@ -61,6 +61,10 @@
+* [Update user group subgroups](/api/update-user-group-subgroups)
+* [Get user group membership](/api/get-user-group-membership)
+* [Get own user group membership](/api/get-own-user-group-membership)
+* [Get user group members](/api/get-user-group-members)
```
I am not sure what will be the best sequence here for the endpoints.

timabbott commented on zerver/lib/user_groups.py:
```diff
@@ -175,6 +175,10 @@ def get_user_group_member_ids(user_group: UserGroup, *, recursive: bool = True)
 
 
+def is_user_in_group(user_group: UserGroup, user: UserProfile, *, recursive: bool = True) -> bool:
+    return user.id in get_user_group_member_ids(user_group, recursive=recursive)
```
I think we'll want to do a more efficient query for this operation, if possible.

sahil839 replied:
I updated the query for this as below -
```python
def is_user_in_group(user_group: UserGroup, user: UserProfile, *, recursive: bool = True) -> bool:
    if not recursive:
        return UserGroupMembership.objects.filter(user_group=user_group, user_profile=user).exists()

    return get_recursive_group_members(user_group=user_group).filter(id=user.id).exists()
```

I am not using `get_user_group_member_ids`, since `get_user_group_direct_members` returns list and it is used in other places so did not want to change it.

timabbott replied:
That works.

timabbott commented on zerver/lib/user_groups.py:
```diff
@@ -10,9 +10,14 @@
         user_group = UserGroup.objects.get(id=user_group_id, realm=user_profile.realm)
+        if for_read:
+            # Everyone is allowed to read a user group and check who are its members.
+            return user_group
```
I think the long-term plan is for each user group to have a "Group of which users can see membership of this user group". 

I think this detail should probably match the current API-level question of whether guests get the direct membership of user groups via the API today. 

But I suppose in the medium term, we may end up with an independent `can_access_user_group_membership()` independent function called both here and in the `register` code path for fetching all user groups. Which in turn will require some way to ask the bulk question with a single database query, i.e. "Fetch the user groups that this user can access".

sahil839 replied:
We currently do not allow guests to access any API endpoint for user groups, even the one that returns all the user groups of the realm and there is no endpoint to get members of a user groups for now. But the guests can get list of all user groups, including the member ids of them, from the `/register` endpoint.

timabbott replied:
Sounds good. I also extended this like so:
```
+        if for_read and not user_profile.is_guest:
+            # Everyone is allowed to read a user group and check who
+            # are its members. Guests should be unable to reach this
+            # code path, since they can't access user groups API
+            # endpoints, but we check for guests here for defense in
+            # depth.
```

I think having that here will be a helpful reminder when we get to changing the groups logic for guests.

timabbott commented:
This is great, thanks for doing this @sahil839! I posted a batch of immediate comments. A few broader notes:
* The API changelog entries seem to be line-wrapped at a much greater width than we normally do; can you clean that up?

sahil839 commented:
@timabbott Updated the PR.
- Updated `is_user_in_group` function as mentioned [here](https://github.com/zulip/zulip/pull/21611#discussion_r839823457).
- Updated the line-wrapping in changelog entries. The ones with long endpoints still have larger widths, like we do in other long endpoints. so that link works correctly.
- Checked the guest access to API endpoints and replied [here](https://github.com/zulip/zulip/pull/21611#discussion_r839699368)
- Started a couple of topics in `#api-design`.

sahil839 commented:
> @timabbott Updated the PR.
> 
> * Updated `is_user_in_group` function as mentioned [here](https://github.com/zulip/zulip/pull/21611#discussion_r839823457).
> * Updated the line-wrapping in changelog entries. The ones with long endpoints still have larger widths, like we do in other long endpoints. so that link works correctly.
> * Checked the guest access to API endpoints and replied [here](https://github.com/zulip/zulip/pull/21611#discussion_r839699368)
> * Started a couple of topics in `#api-design`.

@timabbott this is ready for next review. Changes have been made as per the recent discussion on CZO. Apart from changes mentioned above-
- Updated the endpoints to be accessible only by non-guest users.
- Changed the `recursive` variable name.

timabbott commented on zerver/lib/user_groups.py:
```diff
@@ -26,6 +32,34 @@ def access_user_group_by_id(user_group_id: int, user_profile: UserProfile) -> Us
 
+def get_user_group_from_ids(
+    user_group_ids: Sequence[int], acting_user: UserProfile
+) -> List[UserGroup]:
```
The function name should be plural, and we probably want to call this `access_user_groups_by_id`.

timabbott commented on zerver/lib/user_groups.py:
```diff
@@ -26,6 +32,34 @@ def access_user_group_by_id(user_group_id: int, user_profile: UserProfile) -> Us
+
+    user_group_members_dict: Dict[int, Set[int]] = defaultdict(set)
+    for (user_id, group_id) in user_group_members:
+        user_group_members_dict[group_id].add(user_id)
```
I think this membership check is only going to verify whether the current user is a direct member, right? 

We also probably need to restrict access to system groups here.

timabbott replied:
OK, I think this is the main outstanding issue for this PR. My previous comment is incorrect, given how this function is used. It looks like this function is used to check whether the current user has the ability to access a group for the purpose of making it a member of another group. 

I think probably the rule chosen is reasonable -- basically to add group X to another group, you need to be a moderator or a member of the group. But this would be a good thing to have a #design conversation about unless we see a clear answer for what the rule should be... I don't have a clear idea of what is correct. @alya FYI.

I think we should rename this to `access_user_groups_as_potential_subgroups` or something like that.

timabbott replied:
OK I think after talking with @alya a bit, my proposal is that we only check that this is a group that you can see exists (i.e. that you're not a guest, and they're in your realm, at least with the current model). Neither of us could think of a product reason for why it'd be bad for someone who has permission to add individual users to a group to add another group to that group (even if they have no connection to the new subgroup), which is basically what we're considering here.

Can you post this plan in #design to make sure it gets attention?  I think it's fine to change this function to have that behavior without waiting for that thread to conclude since I think it's unlikely we'll do something different, and `access_user_groups_as_potential_subgroups` seems like a fine name to use.

timabbott commented on zerver/lib/user_groups.py:
```diff
@@ -48,8 +83,16 @@ def user_groups_in_realm_serialized(realm: Realm) -> List[Dict[str, Any]]:
+        "subgroup_id", "supergroup_id"
+    )
+    for (subgroup_id, supergroup_id) in group_membership:
+        group_dicts[supergroup_id]["subgroups"].append(subgroup_id)
```
I'm not sure if there's a way to do this, but ideally we'd do some sort of `select_related` thing that would let us do `UserGroup.objects.select_related("subgroups", "members").filter(realm=realm)` rather than doing 3 database queries here.

timabbott commented on zerver/tests/test_user_groups.py:
```diff
@@ -168,6 +175,25 @@ def test_subgroups_of_role_based_system_groups(self) -> None:
+        self.assertFalse(is_user_in_group(moderators_group, iago, direct_member_only=True))
+
+        self.assertFalse(is_user_in_group(moderators_group, hamlet))
+        self.assertFalse(is_user_in_group(moderators_group, hamlet, direct_member_only=True))
```
I extended this test to have a case where `is_user_in_group(moderators_group, hamlet, direct_member_only=True)` is True.

timabbott commented on zerver/tests/test_user_groups.py:
```diff
@@ -773,3 +799,235 @@ def test_promote_new_full_members(self) -> None:
+        result = self.client_get(f"/json/user_groups/25/members/{iago.id}")
+        self.assert_json_error(result, "Invalid user group")
+
+        result = orjson.loads(
```
Renamed this variable to `result_dict` from here on out, to have consistent types. Generally it'd be better to use an extra line so we can do `assert_json_success(result)`.

timabbott commented on zerver/tests/test_user_groups.py:
```diff
@@ -773,3 +799,235 @@ def test_promote_new_full_members(self) -> None:
+
+        # Invalid user group ID.
+        result = self.client_get(f"/json/user_groups/25/members/{iago.id}")
+        self.assert_json_error(result, "Invalid user group")
```
I suppose ideally these tests would have at least one case where `access_user_group_by_id` or `access_user_by_id` fails due to a permissions reason (E.g. user in a different realm).

I also think we should avoid using a fixed small number like 25; we might eventually have 25 in the test data set. 1000 would be a fine replacement.

These can both be fixed in a follow-up.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -13832,17 +13951,186 @@ paths:
+  /user_groups/{user_group_id}/members/{user_id}:
+    get:
+      operationId: get-user-group-membership
+      summary: Get user group membership
```
This name is too easily mixed up with `get-user-group-members`. I think we want `get-is-user-in-group` or something similar.

timabbott replied:
Changed to `get-is-user-group-member`

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -16179,3 +16476,15 @@ components:
+        Whether to consider only the direct members of user group and not members
+        of its subgroups. Default value is `False`.
+
+        **Changes**: New in Zulip 6.0 (feature level 127).
```
I removed the changes entry, since AFAIK this only appears on endpoints that are new with this feature level.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -16179,3 +16476,15 @@ components:
+      in: query
+      description: |
+        Whether to consider only the direct members of user group and not members
+        of its subgroups. Default value is `False`.
```
@laurynmm should a default be recorded in the OpenAPI metadata another way?

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -13680,6 +13756,41 @@ paths:
+                  - $ref: "#/components/schemas/JsonSuccessBase"
+                  - $ref: "#/components/schemas/SuccessDescription"
+                  - additionalProperties: false
+                    description: MANUALLY
```
What does this do?  It doesn't appear in the end documentation.

sahil839 replied:
I copied this part from an existing endpoint, and it uses this. But I checked and it doesn't appear to be of any use. Should I modify the existing commits only to remove this?

timabbott commented on zerver/lib/user_groups.py:
```diff
@@ -128,6 +171,39 @@ def get_recursive_membership_groups(user_profile: UserProfile) -> "QuerySet[User
+    user_group: UserGroup, *, direct_member_only: bool = False
+) -> List[int]:
+    if direct_member_only:
+        member_ids = get_user_group_direct_members(user_group)
```
This should be a follow-up to avoid excess complexity, but we should probably rename this function to `get_user_group_direct_member_ids`, to avoid it being parallel to `get_recursive_group_members`, which returns a QuerySet.

timabbott commented:
I pushed an updated version of this, addressing most of the minor details noted above. I think we are extremely close to being able to merge it. The main things I'm tracking as blockers are:
* https://github.com/zulip/zulip/pull/21611#discussion_r856717068
* The name for the ``GET /user_groups/{user_group_id}/members/me` endpoint, which I think I mangled while trying to figure out a good name. I think my current thinking is that the best option would be to just have that be a special case of the existing endpoint (i.e. you're allowed to pass "me" for a user ID), rather than an independent endpoint, as I cannot figure out a reasonable name for this and it's not really important. But I'd drop this endpoint from the PR to merge the rest if we get stuck on it.

timabbott commented:
Please do further updates on this PR by adding new commits (possibly ones to be squashed into their parent) -- I think this is sufficiently close that the time savings of my not re-reading many parts that are ready will be well worth it.  Thanks so much for all the work on this @sahil839!

sahil839 commented:
@timabbott updated the PR. Have added two more commits -
- One to address [this comment](https://github.com/zulip/zulip/pull/21611#discussion_r856846275) and this can be squashed to its parent commit.
- One to do a lint fix. Have also kept this as separate commit for now and mentioned the parent commit in commit message. Can squash once you review the PR.

Have not done follow-ups yet, will do in a separate PR. And replied to one comment [here](https://github.com/zulip/zulip/pull/21611#discussion_r857558047).

Regarding the `GET /user_groups/{user_group_id}/members/me` endpoint, I tried merging it with the existing endpoint. Something that we do for `users/<user_id_or_email>/presence` endpoint, where `user_id` is passed as string only, and then parsed in the view function accordingly (as can be seen in `get_presence_backend` [code](https://github.com/zulip/zulip/blob/d0a697fba73b1e0cdfd372b840b8fe4bda225723/zerver/views/presence.py#L28) . But we do have separate endpoints and separate view functions for a couple of other cases like `users/{user_id}` and `users/me` endpoint. If we think having single endpoint is clean and better, we can skip this commit here and add it in separate PR.

timabbott commented:
Cool. A trick when doing these sort of to-squash commits is to have commit messages starting with `fixup!` at the start of the title, and placing them immediately after the commit they're to be squashed into -- that prevents merge conflicts when one tries to actually squash them. And also means one can use https://thoughtbot.com/blog/autosquashing-git-commits.
